browsermt/marian-dev
# Import matrix-multiply from a separate wasm module

### Description

This PR resolves all but last task of the issue https://github.com/browsermt/bergamot-translator/issues/204

List of changes:
For [generic matrix multiply interface](https://github.com/browsermt/marian-dev/blob/master/src/tensors/cpu/wasm_intgemm_interface.h):
- [x] Restructure [marian-dev](https://github.com/browsermt/marian-dev), allowing it to import different implementations of  [generic matrix multiply interface](https://github.com/browsermt/marian-dev/blob/master/src/tensors/cpu/wasm_intgemm_interface.h) for wasm target
- [x] Add a fallback (a polyfill) dummy implementation of this interface in [marian-dev](https://github.com/browsermt/marian-dev) for wasm target and export it from marian wasm module 
    - The fallback dummy implementation can throw initially
- [x] Change emscripten generated JS glue code to plug in this fallback implementation (until an optimized implementation is not landed in Firefox)
- [x] Verify the restructuring by running wasm artifacts on wasm test page
- [x] Provide a valid (non-dummy) fallback implementations for entire [interface](https://github.com/browsermt/marian-dev/blob/master/src/tensors/cpu/wasm_intgemm_interface.h)

**Please note that**:
1. We currently import only those functions that are required to support `--int8shiftAlphaAll` configuration
2. The value `0.0` is being passed as `zero point` everywhere.
5. The wasm target doesn't support 16-bit matrix multiplication (i.e. we are aborting for integer 16-bit matrix multiplication for wasm target).


Added dependencies: none

### How to test
Follow the [README](https://github.com/browsermt/marian-dev/tree/master/wasm#-performing-translation-using-wasm-marian-decoder) instructions to run the wasm test page and observe the translated sentences in browser console logs. This confirms that the matrix-multiply implementation is imported from different wasm modules.

### Checklist
